flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java

```diff
@@ -22,141 +22,153 @@
+	 * reference count reaches <tt>0</tt>.
+	 *
+	 * @see #retain()
+	 */
+	void recycle();
+
+	/**
+	 * Returns whether this buffer has been recycled or not.
+	 *
+	 * @return <tt>true</tt> if already recycled, <tt>false</tt> otherwise
+	 */
+	boolean isRecycled();
+
+	/**
+	 * Retains this buffer for further use, increasing the reference counter by <tt>1</tt>.
+	 *
+	 * @return <tt>this</tt> instance (for chained calls)
+	 *
+	 * @see #recycle()
+	 */
+	Buffer retain();
 
 	/**
-	 * The current size of the buffer in the range from 0 (inclusive) to the
-	 * size of the backing {@link MemorySegment} (inclusive).
+	 * Returns the maximum size of the buffer, i.e. the capacity of the underlying {@link MemorySegment}.
+	 *
+	 * @return size of the buffer
 	 */
-	private int currentSize;
-
-	public Buffer(MemorySegment memorySegment, BufferRecycler recycler) {
-		this(memorySegment, recycler, true);
-	}
-
-	public Buffer(MemorySegment memorySegment, BufferRecycler recycler, boolean isBuffer) {
-		this(memorySegment, recycler, isBuffer, memorySegment.size());
-	}
-
-	public Buffer(MemorySegment memorySegment, BufferRecycler recycler, boolean isBuffer, int size) {
-		this.memorySegment = checkNotNull(memorySegment);
-		this.recycler = checkNotNull(recycler);
-		this.isBuffer = isBuffer;
-		this.currentSize = size;
-	}
-
-	public boolean isBuffer() {
-		return isBuffer;
-	}
-
-	public void tagAsEvent() {
-		synchronized (recycleLock) {
-			ensureNotRecycled();
-		}
-
-		isBuffer = false;
-	}
-
-	public MemorySegment getMemorySegment() {
-		synchronized (recycleLock) {
-			ensureNotRecycled();
-
-			return memorySegment;
-		}
-	}
-
-	public ByteBuffer getNioBuffer() {
-		synchronized (recycleLock) {
-			ensureNotRecycled();
-			// the memory segment returns a distinct buffer every time
-			return memorySegment.wrap(0, currentSize);
-		}
-	}
-	
-	public BufferRecycler getRecycler(){
-		return recycler;
-	}
+	int getMaxCapacity();
 
 	/**
-	 * Returns the size of this buffer without synchronization.
+	 * Returns the <tt>reader index</tt> of this buffer.
+	 *
+	 * <p>This is where readable (unconsumed) bytes start in the backing memory segment.
+	 *
+	 * @return reader index (from 0 (inclusive) to the size of the backing {@link MemorySegment}
+	 * (inclusive))
+	 */
+	int getReaderIndex();
+
+	/**
+	 * Sets the <tt>reader index</tt> of this buffer.
+	 *
+	 * @throws IndexOutOfBoundsException
+	 * 		if the index is less than <tt>0</tt> or greater than {@link #getSize()}
+	 */
+	void setReaderIndex(int readerIndex) throws IndexOutOfBoundsException;
+
+	/**
+	 * Returns the size of the written data, i.e. the <tt>writer index</tt>, of this buffer in an
+	 * non-synchronized fashion.
+	 *
+	 * <p>This is where writable bytes start in the backing memory segment.
+	 *
+	 * @return writer index (from 0 (inclusive) to the size of the backing {@link MemorySegment}
+	 * (inclusive))
+	 */
+	int getSizeUnsafe();
```

Comment:
nit: now I have some doubts about `getSizeUnsafe()`. We should drop it and replace by `getSize()` itself, but maybe that could/should be done in separate PR?

Reply:
I kept it for now